File: python/sieve/sieve.py
def primes(limit):
    numbers = [(0, False), (1, False)] + [(n, None) for n in range(2, limit + 1)]

    while True:
        possibles = (n for (n, status) in numbers if status is None)
        next_prime = next(possibles, None)
        if next_prime is None:
            break

        numbers[next_prime] = (next_prime, True)
        for n in range(next_prime + 1, limit + 1):
            if n % next_prime == 0:
                numbers[n] = (n, False)
    return [n for (n, status) in numbers if status == True]

File: r/sieve/sieve.R
sieve <- function(limit) {
  if (limit < 2) return(c())
  
  numbers <- 2:limit
  
  # status is 1 for prime, -1 for non-prime, 0 for untested
  status <- rep(0, times = limit - 1)

  repeat {
    possibles <- numbers[status == 0]
    
    # exit loop when no numbers left to test
    if (length(possibles) == 0) break
    
    next_prime <- possibles[1]
    status[numbers %% next_prime == 0] <- -1
    status[next_prime - 1] <- 1
  }
  numbers[status == 1]
}
